src/serac/physics/tests/serac_solid_functional_shape.cpp

```diff
@@ -0,0 +1,184 @@
+    // Construct a functional-based solid mechanics solver including references to the shape velocity field.
+    SolidFunctional<p, dim> solid_solver(options, GeometricNonlinearities::On, "solid_functional",
+                                         ShapeDisplacement::On);
+
+    // Set the initial displacement and boundary condition
+    solid_solver.setDisplacementBCs(ess_bdr, bc);
+    solid_solver.setDisplacement(bc);
+
+    solid_solver.shapeDisplacement() = user_defined_shape_displacement;
+
+    solid_solver.setMaterial(mat);
+
+    solid_solver.addBodyForce(force);
+
+    // Finalize the data structures
+    solid_solver.completeSetup();
+
+    // Perform the quasi-static solve
+    double dt = 1.0;
+    solid_solver.advanceTimestep(dt);
+
+    shape_displacement = solid_solver.displacement().gridFunction();
+  }
+
+  axom::sidre::DataStore new_datastore;
+  StateManager::reset();
+  serac::StateManager::initialize(new_datastore, "solid_functional_pure_solve");
+
+  auto new_mesh = mesh::refineAndDistribute(buildMeshFromFile(filename), serial_refinement, parallel_refinement);
+  serac::StateManager::setMesh(std::move(new_mesh));
+
+  {
+    // Construct and initialized the user-defined shape velocity to offset the computational mesh
+    FiniteElementState user_defined_shape_displacement(StateManager::newState(
+        FiniteElementState::Options{.order = p, .vector_dim = dim, .name = "parameterized_shape"}));
+
+    user_defined_shape_displacement.project(shape_coef);
+
+    // Delete the pre-computed geometry factors as we are mutating the mesh
+    StateManager::mesh().DeleteGeometricFactors();
+    auto* mesh_nodes = StateManager::mesh().GetNodes();
+    *mesh_nodes += user_defined_shape_displacement.gridFunction();
+
+    // Construct a functional-based solid mechanics solver including references to the shape velocity field.
+    SolidFunctional<p, dim> solid_solver_no_shape(options, GeometricNonlinearities::On, "solid_functional");
+
+    mfem::VisItDataCollection visit_dc("pure_version", const_cast<mfem::ParMesh*>(&solid_solver_no_shape.mesh()));
+    visit_dc.RegisterField("displacement", &solid_solver_no_shape.displacement().gridFunction());
+    visit_dc.Save();
+
+    // Set the initial displacement and boundary condition
+    solid_solver_no_shape.setDisplacementBCs(ess_bdr, bc_pure);
+    solid_solver_no_shape.setDisplacement(bc_pure);
+
+    solid_solver_no_shape.setMaterial(mat);
+
+    solid_solver_no_shape.addBodyForce(force);
+
+    // Finalize the data structures
+    solid_solver_no_shape.completeSetup();
+
+    // Perform the quasi-static solve
+    double dt = 1.0;
+    solid_solver_no_shape.advanceTimestep(dt);
+
+    pure_displacement = solid_solver_no_shape.displacement().gridFunction();
+    visit_dc.SetCycle(1);
+    visit_dc.Save();
+  }
+
+  for (int i = 0; i < shape_displacement.Size(); ++i) {
+    // Check the final displacement norm
+    EXPECT_NEAR(shape_displacement[i], pure_displacement[i], 1.0e-11);
```

Comment:
👍

Reply:
This is looking better. I would have expected better agreement between these two problems, but we can investigate that later.